### PR TITLE
Remove deprecated code

### LINKED ISSUE #9465: [DEPRECATION] Remove support for boolean masks in take
The `take` method of various pandas objects is only designed to accept integer indices. We have also supported boolean masks up to now, but the preferred method to do this should be via `iloc[mask]`. Removing this helps us reduce the internal complexity of `take`.

### PR DESCRIPTION
**Pull request by vyasr**
This PR removes a large number of deprecated code paths in cuDF. This PR resolves #9465 and partially addresses #9828 (this PR does not address any mask-related API deprecations other than the removal of the already deprecated Series.set_mask).

### CODE REVIEW

File: python/cudf/cudf/core/column/categorical.py [hunk 1]
```diff
@@ -946,7 +949,7 @@ def to_pandas(self, index: pd.Index = None, **kwargs) -> pd.Series:
             col = self
 
         signed_dtype = min_signed_type(len(col.categories))
-        codes = col.codes.astype(signed_dtype).fillna(-1).to_array()
+        codes = col.codes.astype(signed_dtype).fillna(-1).values_host
```
Comment (bdice) [inline python/cudf/cudf/core/column/categorical.py:952]: Is this `-1` used to mean the same thing as `_DEFAULT_CATEGORICAL_VALUE`?
```suggestion
        codes = col.codes.astype(signed_dtype).fillna(_DEFAULT_CATEGORICAL_VALUE).values_host
```
Same question here: https://github.com/rapidsai/cudf/blob/f46c423f18a26aef8db3bf62b3e6b10eed015884/python/cudf/cudf/core/column/categorical.py#L1604

File: python/cudf/cudf/core/column/datetime.py [hunk 2]
```diff
@@ -199,7 +199,7 @@ def to_pandas(
 
         # Pandas supports only `datetime64[ns]`, hence the cast.
         return pd.Series(
-            self.astype("datetime64[ns]").to_array("NAT"),
+            self.astype("datetime64[ns]").fillna("NAT").values_host,
```
Comment (bdice) [inline python/cudf/cudf/core/column/datetime.py:202]: Prefer the capitalizations `NaN` and `NaT` for consistency with their standard pandas representations.
```suggestion
            self.astype("datetime64[ns]").fillna("NaT").values_host,
```

File: python/cudf/cudf/core/column/timedelta.py [hunk 3]
```diff
@@ -123,7 +117,8 @@ def to_pandas(
 
         # Pandas supports only `timedelta64[ns]`, hence the cast.
         pd_series = pd.Series(
-            self.astype("timedelta64[ns]").to_array("NAT"), copy=False
+            self.astype("timedelta64[ns]").fillna("NAT").values_host,
```
Comment (bdice) [inline python/cudf/cudf/core/column/timedelta.py:120]:
```suggestion
            self.astype("timedelta64[ns]").fillna("NaT").values_host,
```

File: python/cudf/cudf/testing/_utils.py [hunk 4]
```diff
@@ -40,6 +40,9 @@
 ALL_TYPES = sorted(list(dtypeutils.ALL_TYPES))
 
 
+# TODO: This function should be removed. Anywhere that it is being used should
+# instead be generating a random boolean array (bytemask) and use the public
+# APIs to set those elements to None.
```
Comment (bdice) [inline python/cudf/cudf/testing/_utils.py:45]: I have done some related (unfinished) work in this PR. I need to do some thinking on how to best refactor that draft for performance and readability. https://github.com/rapidsai/cudf/pull/10034

**Review by bdice — CHANGES_REQUESTED** [review]
Very nice! I have a few comments/questions.

**Comment by beckernick** [comment]
+34, -471 is beautiful

File: python/cudf/cudf/core/column/categorical.py (hunk 1 above)
Comment (vyasr) [reply python/cudf/cudf/core/column/categorical.py:952]: Good catch, I did a find and replace on the sensible places.

File: python/cudf/cudf/testing/_utils.py (hunk 4 above)
Comment (vyasr) [reply python/cudf/cudf/testing/_utils.py:45]: Ah my apologies, I did not see that PR (perhaps since it's a draft?). I didn't mean to step on your toes there. IMO tests shouldn't rely on internal APIs any more than users should, and the way for a user to apply a mask is via a bytemask `series[bytemask] = None`, so I think it's fine for tests to also use that approach even if it is less efficient. I think this PR can remove `set_mask` but change the tests to use the internal column APIs (which is what I've done so far), and then you can make the subsequent changes to the tests in #10034 once we've come to a consensus (in case you disagree with my assertion above). Sound good?

**Comment by vyasr** [comment]
> +34, -471 is beautiful

Now +99, -687 :)

File: python/cudf/cudf/testing/_utils.py (hunk 4 above)
Comment (bdice) [reply python/cudf/cudf/testing/_utils.py:45]: No worries! I started that draft but it was low priority. It might be helpful to have a cudf.testing function to assign a random null mask, since that pattern is repeated so many times in the tests. I agree that we should avoid internal APIs in tests to the greatest extent possible.

File: python/cudf/cudf/tests/test_parquet.py [hunk 5]
```diff
@@ -2124,7 +2123,7 @@ def test_parquet_writer_statistics(tmpdir, pdf, add_nulls):
     gdf = cudf.from_pandas(pdf)
     if add_nulls:
         for col in gdf:
-            gdf[col] = gdf[col].set_mask(random_bitmask(len(gdf)))
+            gdf[col][np.random.choice([False, True], size=len(gdf))] = None
```
Comment (bdice) [inline python/cudf/cudf/tests/test_parquet.py:2126]: My biggest regret so far in the design of draft PR #10034 was that I hadn't yet introduced a `cudf.testing` function to do this, and found myself re-implementing this ugly numpy logic many times over. I think it's hard to read (hard to parse by eye and requires some nontrivial understanding of `np.random.choice`). Thus, I'd like to give it a name, even though it's only one line of code. Something like this would be my suggestion:
```python
def set_random_null_mask_inplace(series, null_probability=0.5):
    probs = [null_probability, 1-null_probability]
    series[np.random.choice([False, True], size=len(series), p=probs)] = None
```

Comment (vyasr) [reply python/cudf/cudf/tests/test_parquet.py]: I was going to wait to do this in a future PR, but since you requested it I went ahead and added a function. I didn't change its usage anywhere except here though; I'll leave that to you in #10034.

Comment (bdice) [reply python/cudf/cudf/tests/test_parquet.py]: I really like what you did in c01d0f2! Approving.

**Review by bdice — APPROVED** [review]
Nice work. I'll follow up on this in #10034 to use the new `set_random_null_mask_inplace(sr)` helper function in all tests. I think this PR is complete in its intended scope so I approve.

**Review by shwina — APPROVED** [review]

**Comment by shwina** [comment]
@gpucibot merge

**Comment by bdice** [comment]
I'll pick up #10034 today / Monday. Thanks again @vyasr!
